### PR TITLE
[contracts] Test fillOrder return values and abiEncodeFillOrder

### PR DESCRIPTION
**Pull request by abandeali1**
## Description

- Test `FillResults` and `abiEncodeFillOrder` within the combinatorial tests for `fillOrder`
- Increment the free memory pointer at the end of the assembly blocks in `dispatchTransferFrom` and `abiEncodeFillOrder`. It _shouldn't_ be an issue to overwrite this memory (which was happening previously), but feels like a good safety addition.

## Checklist:

*   [ ] Prefixed the title of this PR with `[WIP]` if it is a work in progress.
*   [ ] Prefixed the title of this PR with bracketed package name(s) corresponding to the changed package(s). For example: `[sol-cov] Fixed bug`.
*   [ ] Added tests to cover my changes, or decided that tests would be too impractical.
*   [ ] Updated documentation, or decided that no doc change is needed.
*   [ ] Added new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

**Comment by abandeali1** [comment]
Looks like incrementing the free memory pointer adds ~100 gas per fill due to memory expansion. @Recmo @hysz thoughts on if this is a good tradeoff?

File: packages/contracts/src/2.0.0/protocol/Exchange/MixinAssetProxyDispatcher.sol [hunk 1]
```diff
@@ -159,18 +159,22 @@ contract MixinAssetProxyDispatcher is
...
+                    inputLen,               // length of input  
                     cdStart,                // write output over input
                     512                     // reserve 512 bytes for output
                 )
                 if iszero(success) {
                     revert(cdStart, returndatasize())
                 }
+
+                // Increment free memory pointer
+                mstore(64, add(cdStart, inputLen))
```
Comment (Recmo) [inline packages/contracts/src/2.0.0/protocol/Exchange/MixinAssetProxyDispatcher.sol]: The memory is discarded within the asm block, so we don't need to increment the free memory pointer. (And if we do it should have been the max of `inputLen` and `returndatasize()` :) )

Note that Solidity does the same thing when it generates assembly for a call.

Gas cost will get *quadratically* worse for large batch orders.

File: packages/contracts/src/2.0.0/protocol/Exchange/libs/LibAbiEncoder.sol [hunk 2]
```diff
@@ -207,10 +204,11 @@ contract LibAbiEncoder {
...
             // Set length of calldata
-            mstore(
-                fillOrderCalldata,
-                sub(dataAreaEnd, add(fillOrderCalldata, 0x20))
-            )
+            let calldataLen := sub(dataAreaEnd, add(fillOrderCalldata, 0x20))
+            mstore(fillOrderCalldata, calldataLen)
+
+            // Increment free memory pointer
+            mstore(0x40, add(fillOrderCalldata, add(calldataLen, 0x20)))
```
Comment (Recmo) [inline packages/contracts/src/2.0.0/protocol/Exchange/libs/LibAbiEncoder.sol]: This looks fine, but `mstore(0x40, dataAreaEnd)` should also work.

File: packages/contracts/src/2.0.0/protocol/Exchange/libs/LibAbiEncoder.sol [hunk 3]
```diff
@@ -207,10 +204,10 @@ contract LibAbiEncoder {
...
 
             // Set length of calldata
-            mstore(
-                fillOrderCalldata,
-                sub(dataAreaEnd, add(fillOrderCalldata, 0x20))
-            )
+            mstore(fillOrderCalldata, sub(dataAreaEnd, add(fillOrderCalldata, 0x20)))
+
+            // Increment free memory pointer
+            mstore(0x40, add(fillOrderCalldata, dataAreaEnd))
```
Comment (Recmo) [inline packages/contracts/src/2.0.0/protocol/Exchange/libs/LibAbiEncoder.sol]: This looks off. I think it should be `mstore(0x40, dataAreaEnd)`. After the signature loop, `dataAreaEnd` will point to the first byte beyond the last one written. Adding `fillOrderCalldata` would add two pointers, which is weird.

Comment (abandeali1) [reply packages/contracts/src/2.0.0/protocol/Exchange/libs/LibAbiEncoder.sol]: Whoops, that is most definitely a mistake

**Pull request merged**